ddl/ddl_api.go

```diff
@@ -2101,10 +2100,18 @@ func (d *ddl) AddColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTab
 			if err = checkAutoIncrementRef(specNewColumn.Name.Name.L, dependColNames, t.Meta()); err != nil {
 				return errors.Trace(err)
 			}
+			duplicateColNames := make(map[string]struct{})
+			for k := range dependColNames {
+				duplicateColNames[k] = struct{}{}
+			}
 
 			if err = checkDependedColExist(dependColNames, t.Cols()); err != nil {
 				return errors.Trace(err)
 			}
+
+			if err = verifyColumnGenerationSingle(duplicateColNames, t.Cols(), spec.Position); err != nil {
```

Comment:
Please extract `t.Cols()` in this line and line 2108, avoiding unnecessary function call.

Reply:
addressed.